stan/math/prim/scal/fun/F32.hpp

```diff
@@ -1,42 +1,89 @@
 #ifndef STAN_MATH_PRIM_SCAL_FUN_F32_HPP
 #define STAN_MATH_PRIM_SCAL_FUN_F32_HPP
 
+#include <stan/math/prim/scal/fun/sign.hpp>
+#include <stan/math/prim/scal/err/domain_error.hpp>
+#include <stan/math/prim/scal/fun/is_nan.hpp>
+#include <stan/math/prim/scal/fun/is_inf.hpp>
 #include <cmath>
 
 namespace stan {
   namespace math {
 
+    /**
+     * Hypergeometric function, 3F2.
+     *
+     * The generalized hypergeometric function is a power series. This
+     * implementation computes the power series directly stopping when
+     * the series converges to within <code>precision</code> or takes
+     * <code>max_steps</code>.
+     *
+     * Although some convergence conditions and divergent conditions are known,
+     * this function does not check the inputs for known convergent conditions.
+     *
+     * This function will converge if:
+     *   - <code>a1</code>, <code>a2</code>, or <code>a3</code> is a
+     *     non-positive integer
+     *   - <code>b1</code> or <code>b2</code> is a non-positive integer
+     *   - <code>z</code> is less than 1
+     * This function will diverge if <code>z</code> is greater than 1.
+     * When <code>z</code> is 1, which is the case for the beta binomial
+     * cdf, it is hard to determine.
+     *
+     * @tparam T type of arguments and result
+     * @param[in] a1 a1 (always called with 1 from beta binomial cdfs)
+     * @param[in] a2 a2 (always called with a2 > 1)
+     * @param[in] a3 a3 (always called with int a3 <= 0)
+     * @param[in] b1 b1 (always called with int b1 < |a3|)
+     * @param[in] b2 b2 (always <= 1)
+     * @param[in] z z (is always called with 1 from beta binomial cdfs)
+     * @param[in] precision precision of the infinite sum. defaults to 1e-6
+     * @param[in] max_steps number of steps to take. defaults to 10000
+     */
     template<typename T>
-    T F32(T a, T b, T c, T d, T e, T z, T precision = 1e-6) {
+    T F32(const T& a1, const T& a2, const T& a3, const T& b1, const T& b2,
+          const T& z, double precision = 1e-6, int max_steps = 1e5) {
       using std::exp;
       using std::log;
       using std::fabs;
+      using stan::math::is_nan;
 
       T F = 1.0;
-
       T tNew = 0.0;
-
       T logT = 0.0;
-
       T logZ = log(z);
 
-      int k = 0.0;
-
-      while (fabs(tNew) > precision || k == 0) {
-        T p = (a + k) * (b + k) * (c + k) / ( (d + k) * (e + k) * (k + 1) );
-
-        // If a, b, or c is a negative integer then the series terminates
-        // after a finite number of interations
-        if (p == 0) break;
-
-        logT +=  (p > 0 ? 1.0 : -1.0) * log(fabs(p)) + logZ;
+      int k = 0;
+      bool T_is_negative = false;
+      T p = 0.0;
+      do {
+        p = (a1 + k) / (b1 + k) * (a2 + k) / (b2 + k) * (a3 + k) / (k + 1);
 
-        tNew = exp(logT);
+        if (is_nan(p) || p == 0)
```

Comment:
@betanalpha many sources seem to say the function is undefined when any denominator is a negative integer, mathematica gives an answer indistinguishable from terminating (as above) at the last defined term of the sum.  For this PR I'll either leave the code matching Mathematica or throw an error, I'll leave improving this situation for issue #511 as the current implementation doesn't do any better than this. Sound ok?